### ISSUE TITLE
Training with deep residual network

### ISSUE THREAD

**Issue by 201power** [KaimingHe/deep-residual-networks#6]
@KaimingHe Thanks for sharing the network. 

I've modified your 50 layer resnet deploy prototxt for training (8 GPU, 128 batch size with 0.01 learning rate), however, it's not converge, can you suggest any possible causes? Thanks. 

I added initial parameter for convolution and scale layers, example:

**Convolution Layer**

```
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  convolution_param {
    num_output: 64
    pad: 3
    kernel_size: 7
    stride: 2
    bias_term: false
    weight_filler {
      type: "msra"
    }
  }
```

**Scale Layer**

```
layer {
  name: "conv1_scale"
  type: "Scale"
  bottom: "conv1"
  top: "conv1"
  param {
    lr_mult: 2
    decay_mult: 0
  }	
  scale_param {
    bias_term: true
    bias_filler {
      type: "constant"
      value: 0
    }
  }
```
I train the resnet on my own datasets (6000 images), which converge on Googlenet / Alexnet. 

![image](https://cloud.githubusercontent.com/assets/7299296/12988054/d1684740-d0b3-11e5-8dc2-27e88f4cda4b.png)

**Comment by KaimingHe** [comment]
@201power I cannot diagnose from what you provided. Please check Facebook's implementation (https://github.com/facebook/fb.resnet.torch) for the hyper-parameters. ResNets could converge within a rather wide range of hyper-parameters.

**Comment by beniz** [comment]
@201power check out https://github.com/beniz/deepdetect/tree/master/templates/caffe/resnet_50 as my modified `prototxt` appear to converge on a range of datasets (smaller than Imagenet).

**Comment by 201power** [comment]
@KaimingHe Thanks. 
@beniz I tried your network and it still does not converge on my data. I think there is an issue with my caffe and since I manually added the scale and bias layer to the nvidia caffe.

**Comment by beniz** [comment]
@201power try lower learning rate if you haven't already, from a range of experiments, 0.001 and below did work fine and straight out of the box with dd.

**Comment by beniz** [comment]
@201power Last report here for me, so you can have a reference basis: training resnet-50 starting with learning rate of 0.01 and ADAM solver converges nicely on an 18K images problem. Good luck.

**Comment by lfrdm** [comment]
@201power I think the problem lies with the correct usage of batch normalization. One must differentiate between training and testing. Have a look at my <a href="https://github.com/lfrdm/Masterarbeit/blob/master/train_val-residual-small.prototxt">prototxt</a>. !! NOTE !! that I decreased all outputs by a factor of 4 for run time purposes and only a 2 class problem.

**Comment by 201power** [comment]
@beniz, I tried 0.01 or 0.001 and it still did not work. I'll try ADAM solver. 
@lfrdm, why do you need decrease outputs by a factor of 4?

**Comment by lfrdm** [comment]
@201power the original residual net is designed for the 1000 class dataset of imagenet and other complex classification tasks. I use it for classification of only 2 classes. Furthermore, it trains much faster and takes less memory with a smaller number of feature maps.

**Comment by siddharthm83** [comment]
@beniz Have you tried training ResNet on Imagenet-1000 from scratch? I used your deploy and trainval and my loss is not converging. Any tips?

**Comment by beniz** [comment]
@siddharthm83 not on ILSVRC yet, but `resnet_50` with deepdetect on a range of smaller problems is converging nicely. If you are using deepdetect as well, you can open an issue and it will be looked upon sometime. I cannot offer support for other frameworks.

**Comment by n3011** [comment]
hi @beniz I have tried your prototxt fro resnet_50, but its not converging? Any idea?

**Comment by beniz** [comment]
@n3011 as stated above, please see https://github.com/beniz/deepdetect/issues/84 as we cannot offer support outside of dd. The issue already contains some useful tips and feedback. If you decide to join the conversation, please state training calls, exact parameters and whether you are using a public dataset so that the issue can be reproduced.

**Comment by beniz** [comment]
@KaimingHe sorry about the noise to this thread related to our external set of files, definitely not the primary intention.

**Comment by lufo816** [comment]
I meet the same problem too, but after i set batch_norm_param:use_global_stats to false all goes well. @201power

**Comment by KeyKy** [comment]
@beniz I saw your deepdetect prototxt of resnet_50.prototxt. According to caffe batch_norm_layer.hpp, you MUST set the learning rate to zero for all three parameter blobs, i.e. param {lr_mult: 0} three times in the layer definition. However, I found that you did not set the lr_mult: 0 in your batch_norm_layer.

**Comment by beniz** [comment]
@KeyKy thanks that's a good catch, however, I've been able to train and finetune resnets without issues for months. When looking at `net.cpp`, `lr_mult==0` seems to avoid the backward pass, which I'm not sure we want here.

In all cases, I suggest you try adding

```
 param {
    lr_mult: 0
  }
  param {
    lr_mult: 0
  }
  param {
    lr_mult: 0
  }
```

to every `BatchNorm` layer, and report whether you see better/good results and convergence. You can also open a dedicated issue on DD directly, and it will be considered.

**Comment by pfollmann** [comment]
I had similar issues using the bvlc caffe version. The problem was with an error inside the BatchNorm layer. I put a pull request here https://github.com/BVLC/caffe/pull/5136
